src/util/opaque.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <variant>
+
+namespace bpftrace::util {
+
+// OpaqueValue corresponds to a blob of type-erased memory. It owns its own
+// underlying memory, taking a copy if necessary. This allows it to be used
+// to read maps, rings, and other locations which may not be permanent.
+//
+// This class does not really provide safety beyond first-order memory safety.
+class OpaqueValue {
+private:
+  struct OwnedBuffer {
+    OwnedBuffer(char *data) : data(data) {};
+    ~OwnedBuffer()
+    {
+      free(data);
+    }
+    char *data;
+  };
+  using SharedMemory = std::shared_ptr<OwnedBuffer>;
+  using Variant = std::variant<uintptr_t, SharedMemory>;
+
+public:
+  // Allows for the creation of an OpaqueValue.
+  static OpaqueValue alloc(size_t length, std::function<void(char *)> init)
+  {
+    return { length, init };
+  }
+
+  // Create a copy of a string.
+  static OpaqueValue string(const std::string &s, size_t sz)
+  {
+    return { sz, [&](char *data) {
+              size_t s_sz = std::min(s.size() + 1, sz);
+              memcpy(data, s.data(), s_sz);
+              memset(data + s_sz, 0, sz - s_sz);
+            } };
+  }
+
+  // Merge two OpaqueValues together.
+  OpaqueValue operator|(const OpaqueValue &other) const
+  {
+    return { size() + other.size(), [&](char *new_data) {
+              memcpy(new_data, data(), size());
+              memcpy(new_data + size(), other.data(), other.size());
+            } };
+  }
+
+  // Creates a copy of any type that has an iterator.
+  //
+  // This requires that the underlying type has a simple `from`, so we can't do
+  // complex types like strings in this way. This is intended for use in tests
+  // only and is therefore not optimized for performance in any way.
+  template <std::ranges::range T>
+  static OpaqueValue from(const T &other)
+  {
+    OpaqueValue res;
+    for (const auto &v : other) {
+      res = res | from(v);
+    }
+    if constexpr (std::is_same_v<T, std::string>) {
+      res = res | OpaqueValue::from<char>(0);
+    }
+    return res;
+  }
+
+  // Creates an OpaqueValue from an existing region.
+  template <typename T>
+  static OpaqueValue from(const T *orig_data, size_t length)
+  {
+    return OpaqueValue(sizeof(T) * length, [&](char *data) {
+      memcpy(data,
+             reinterpret_cast<const char *>(orig_data),
+             sizeof(T) * length);
+    });
```

Comment:
Use the same style as elsewhere?
```suggestion
    return { sizeof(T) * length, [&](char *data) {
             memcpy(data,
                    reinterpret_cast<const char *>(orig_data),
                    sizeof(T) * length);
           } };
```